aten/src/ATen/core/ivalue.cpp

```diff
@@ -604,4 +604,45 @@ getClassConverter() {
       classConverter;
   return classConverter;
 }
+
+CAFFE2_API intrusive_ptr<ivalue::Future> collectAll(
+    List<intrusive_ptr<ivalue::Future>> srcs) {
+  struct Ctx {
+    explicit Ctx(List<intrusive_ptr<ivalue::Future>> srcs)
+        : remaining(srcs.size()),
+          srcFutures(std::move(srcs)),
+          asIvalue(srcFutures),
+          dstFuture(make_intrusive<ivalue::Future>(asIvalue.type())) {}
+    std::atomic<int32_t> remaining{0};
+    List<intrusive_ptr<ivalue::Future>> srcFutures;
+    IValue asIvalue;
+    intrusive_ptr<ivalue::Future> dstFuture;
+  };
+
+  auto ctx = std::make_shared<Ctx>(std::move(srcs));
+  std::function<void()> func = [ctx]() {
+    if (--ctx->remaining == 0) {
+      ctx->dstFuture->markCompleted(ctx->asIvalue);
+    }
+  };
+  for (int32_t tot = ctx->srcFutures.size(), i = 0; i < tot; ++i) {
+    ctx->srcFutures.get(i)->addCallback(func);
+  }
+  if (ctx->srcFutures.size() == 0) {
+    ctx->dstFuture->markCompleted(ctx->asIvalue);
```

Comment:
The IValue type of the `dstFuture` is a list of futures? And this is because we wanted to conveniently retrieve all results from the `dstFuture`?

Reply:
Not sure if there's a better suggestion?

It seems that the result should be a Future of some sort. It seems like the second-best alternative would be a Future<Null> where the called needs to remember the passed in futures out-of-band? But this didn't necessarily seem better...

A `Future<List<IValue>>` isn't really possible because of the need to propagate exceptions back. which is why it was effectively `Future<List<Future<IValue>>>`

Reply:
I see. This current version LGTM.